File: facerecognization/android/app/src/main/java/com/attendance/domain/sync/SyncWorker.kt
package com.attendance.domain.sync

import android.content.Context
import androidx.work.CoroutineWorker
import androidx.work.WorkerParameters
import com.attendance.data.local.AppDatabase
import com.attendance.data.remote.ApiService
import com.attendance.data.remote.SyncRequest
import com.attendance.data.remote.OfflineLogPayload
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class SyncWorker(
    context: Context,
    params: WorkerParameters
) : CoroutineWorker(context, params) {

    override suspend fun doWork(): Result {
        val database = AppDatabase.getDatabase(applicationContext)
        val unsyncedLogs = database.attendanceLogDao().getUnsyncedLogs()

        if (unsyncedLogs.isEmpty()) {
            return Result.success()
        }

        // Load configuration from local preferences
        val sharedPrefs = applicationContext.getSharedPreferences("attendance_prefs", Context.MODE_PRIVATE)
        val hostUrl = sharedPrefs.getString("host_url", null) ?: return Result.failure()
        val token = sharedPrefs.getString("access_token", null) ?: return Result.failure()

        val apiService = Retrofit.Builder()
            .baseUrl(hostUrl)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(ApiService::class.java)

        val payload = SyncRequest(
            logs = unsyncedLogs.map {
                OfflineLogPayload(
                    employee_id = it.employeeId,
                    clock_time = it.clockTime,
                    clock_type = it.clockType,
                    status = it.status,
                    device_id = it.deviceId
                )
            }
        )

        return try {
            val response = apiService.syncLogs("Bearer $token", payload)
            if (response.isSuccessful) {
                // Update Room entries flag state
                unsyncedLogs.forEach {
                    database.attendanceLogDao().markSynced(it.id)
                }
                Result.success()
            } else {
                Result.retry()
            }
        } catch (e: Exception) {
            Result.retry()
        }
    }
}

File: facerecognization/flutter_client/lib/domain/sync_worker.dart
import 'dart:convert';
import 'package:workmanager/workmanager.dart';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import '../data/local_db.dart';

void callbackDispatcher() {
  Workmanager().executeTask((task, inputData) async {
    final localDb = LocalDatabase.instance;
    final unsyncedLogs = await localDb.getUnsyncedLogs();

    if (unsyncedLogs.isEmpty) {
      return Future.value(true);
    }

    final prefs = await SharedPreferences.getInstance();
    final host = prefs.getString('host_url');
    final token = prefs.getString('access_token');

    if (host == null || token == null) {
      return Future.value(false);
    }

    final payload = unsyncedLogs.map((log) => {
      'employee_id': log['employeeId'],
      'clock_time': log['clockTime'],
      'clock_type': log['clockType'],
      'status': log['status'],
      'device_id': log['deviceId']
    }).toList();

    try {
      final response = await http.post(
        Uri.parse('$host/api/v1/attendance/sync'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token'
        },
        body: jsonEncode({'logs': payload}),
      );

      if (response.statusCode == 200) {
        for (var log in unsyncedLogs) {
          await localDb.markSynced(log['id'] as int);
        }
        return Future.value(true);
      }
      return Future.value(false);
    } catch (e) {
      return Future.value(false);
    }
  });
}
